### PR TITLE
[AP-4035] Enhance file upload with dropzone

### PR DESCRIPTION
**Pull request by jsugarman**
## What
Enhance file upload functionality

[Link to story](https://dsdmoj.atlassian.net/browse/AP-4035)

## Checklist

Before you ask people to review this PR:

- Tests and linters should be passing: `bundle exec rake`
- Github should not be reporting conflicts; you should have recently run `git rebase main`.
- There should be no unnecessary whitespace changes. These make diffs harder to read and conflicts more likely.
- The PR description should say what you changed and why, with a link to the JIRA story.
- You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: app/javascript/src/dropzone/bulk_submission_form_controller.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+// the `uploaded_file`` attribute (specified by paramName). Therefore the
+// `error` event handles both these possible response types.
+//
+import Dropzone from 'dropzone'
+
+import {
+  addErrorMessage,
+  removeErrorMessages
+} from './govuk_errors_helper'
+
```
Comment (RoseSAK) [inline app/javascript/src/dropzone/bulk_submission_form_controller.js:20]: This is a good idea separating out all the error stuff 👍  we should probably do this for the apply dropzone too

File: app/views/bulk_submission_forms/_form_fields.html.erb [hunk 2]
```diff
@@ -1,26 +1,65 @@
...
+  <!-- Dropzone upload form -->
+  <div class="dropzone__upload">
+    <div class="govuk-form-group script hidden <%= form_error_class %>" id="dropzone-form-group" aria-labelledby="dropzone-label">
+      <label class="govuk-label govuk-label--m" id="dropzone-label">
+          <%= t("generic.attach_file") %>
+      </label>
+
+      <div class="govuk-hint"><%= t(".size_hint") %></div>
+
+      <% if form.object.errors %>
```
Comment (RoseSAK) [inline app/views/bulk_submission_forms/_form_fields.html.erb:19]: I think you might be able to remove this whole if block - I might be wrong but in the original JS I think this block was added to display errors directly from the form object. But here all the errors from the form object are returned as JSON and rendered via the dropzone error handle, is that right?

**Review by RoseSAK — APPROVED** [review]
Added a couple of comments but on the whole looks great to me 💯

File: app/views/bulk_submission_forms/_form_fields.html.erb (hunk 2 above)
Comment (jsugarman) [reply app/views/bulk_submission_forms/_form_fields.html.erb:19]: thanks, I will take a look. we do pass the errors via JSON so if the summary and field level are being generated via the same JSON then hopefully this can go 🤞

**Review by kmahern — APPROVED** [review]
LGTM

File: app/views/bulk_submission_forms/_form_fields.html.erb (hunk 2 above)
Comment (jsugarman) [reply app/views/bulk_submission_forms/_form_fields.html.erb:19]: Turns out this block is necessary for an "edge" case where a user hits Save and continue without adding any file.

If no file is chosen then, while no-JS will add the "You must select a file" error to the summary _and file field_, the JS/Dropzone file field will not get the error added (because the file field is a dropzone file field that is not found by govuk form builder gem's handling) and **dropzone has not been invoked**, so its `on('error')` event never gets the chance to add it either. So this block patches that edge case.

on the plus side, the system tests caught this edge case explicitly even though my manual testing did not 😄

Comment (RoseSAK) [reply app/views/bulk_submission_forms/_form_fields.html.erb:19]: Ahh okay, makes sense, thanks for investigating/explaining!
